stdlib/src/builtin/constrained.mojo

```diff
@@ -33,3 +33,16 @@ fn constrained[cond: Bool, msg: StringLiteral = "param assertion failed"]():
         cond = cond.__mlir_i1__(), message = msg.value
     ]()
     return
+
+
+@always_inline("nodebug")
+fn unreachable[msg: StringLiteral = "unreachable code"]():
```

Comment:
This is an extremely misleading name for the function. It doesn't actually indicate unreachable code in the control-flow sense, which will lead to lots of misuse of the function. For example,


```mojo
fn make_true() -> Boo: return True

fn infinite_loop():
    while make_true():
        pass
    unreachable()
```

Will fail to compile

Reply:
I think that's more of a compiler edge case than a misnamed function.

Reply:
The compiler can't interprocedurally reason about the function call and prune dead code in order to make the program "correct". That would make the optimization level affect whether code compiles

Reply:
Plus, this function doesn't actually model a divergent function call. Type checking will still complain that a function like

```
fn foo() -> Int:
    unreachable()
```

Doesn't return an `Int` as expected.

Reply:
But when is an example when the above code would be functioning as intended? Why is there an unreachable in the first place?

Reply:
> Plus, this function doesn't actually model a divergent function call. Type checking will still complain that a function like
> 
> 
> 
> ```
> 
> fn foo() -> Int:
> 
>     unreachable()
> 
> ```
> 
> 
> 
> Doesn't return an `Int` as expected.

I'd argue that is a valid error for that function

Reply:
The point is that the name "unreachable" suggests divergent control flow, which is not what the function implements. The function you implemented is a `static_assert(false)`, which is something else.

Reply:
Do you have an alternative name that is more correct, or different implementation?